_posts/2021-06-06-spring-camel.md

```diff
@@ -0,0 +1,422 @@
+
+
+```java
+@Component
+public class FetchProductsRoute extends RouteBuilder {
+
+  @Override
+  public void configure() throws Exception {
+    from("direct:fetchProducts")
+      .routeId("direct-fetchProducts")
+      .tracing()
+      .log(">>> ${body}")
+      .bean(ProductService.class, "fetchProductsByCategory")
+      .end();
+  }
+
+}
+```
+Here we are creating the route by defining the Java DSL in a class `FetchProductsRoute` by extending `RouteBuilder` class. We defined the endpoint as `direct:fetchProducts` and provided a route identifier `direct-fetchProducts`. The prefix `direct:` in the name of the endpoint makes it possible to call the route from another camel route. 
+
+### Triggering a Route with Templates
+We can invoke the routes with `ProducerTemplate` and `ConsumerTemplate`. The `ProducerTemplate` is used as an easy way of sending messages to a Camel endpoint. Both of these templates are similar to the template utility classes in the Spring Framework like `JmsTemplate` or `JdbcTemplate` that simplify access to the JMS and JDBC APIs. 
+
+Let us invoke the route we created earlier from a resource class in our application :
+
+```java
+@RestController
+public class ProductResource {
+  
+  @Autowired
+  private ProducerTemplate producerTemplate;
+  
+  @GetMapping("/products/{category}")
+  @ResponseBody
+  public List<Product> getProductsByCategory(@PathVariable("category") final String category){
+    producerTemplate.start();
+    List<Product> products = producerTemplate
+       .requestBody("direct:fetchProducts", category, List.class);
+    
+    producerTemplate.stop();
+    return products;
+  
+  }
+} 
+
+@Configuration
+public class AppConfig {
+  
+  @Autowired
+  private  CamelContext camelContext;
+ ...
+ ...
+  
+  @Bean
+  ProducerTemplate producerTemplate() {
+    return camelContext.createProducerTemplate();
+  }
+  
+  @Bean
+  ConsumerTemplate consumerTemplate() {
+    return camelContext.createConsumerTemplate();
+  }
+
+}
+
+```
+
+Here we have defined a REST endpoint in our `resource` class with a `GET` method for fetching products by category. We are invoking our Camel route inside the method by using the `producerTemplate` which we configured in our Spring configuration. 
+
+In our Spring configuration we have defined the `producerTemplate` and `consumerTemplate` by calling corresponding methods on the `CamelContext` which is available in the `ApplicationContext`.
+
+## Defining a Route with Splitter-Aggregator Enterprise Integration Pattern
+
+Let us now look at a route where we will use a Enterprise Integration Pattern.
+
+Camel provides implementations for many of the [Enterprise Integration Patterns](https://www.enterpriseintegrationpatterns.com/patterns/messaging/toc.html) from the book by Gregor Hohpe and Bobby Woolf. Here is a snippet of the list of those patterns :
+
+![EIP Snippet](/assets/img/posts/camel-spring/eip-snippet.png)
```

Comment:
This screenshot doesn't really help the reader understand enterprise integration patterns. Instead, maybe explain the Splitter and Aggregator patterns in a sentence each, because these are the ones you're using later?